### ISSUE TITLE
Work on a better workflow (toward CI / CD)

### ISSUE THREAD

**Issue by jooola** [airsonic/airsonic#1561]
## Problem description

An interesting chapter that inspire this ticket https://landing.google.com/sre/sre-book/chapters/eliminating-toil/

The project is getting more and more contribution and we currently have a few problems regarding the workflow. Some concerns are cross repository, meaning we need to take into count the website and documentation repository as well. 

- Release process was at first in the hands of @muff1nman, has been documented in the a RELEASE documentation file, but hasn't reach the step of been integrated in a CI/CD workflow. I see some improvement here.

- Change log could be build automatically, using either [Angular commit message format](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit) or using github labels to sort the kind of the changes that have been made.

- Current testing in Travis seems to be full of false positive, or really unstable. I need some insight on this one, has been a while I've not been around. (Also this could have a separate issue)

- Has checkstyle been integrated in the maven tests ? It this already integrated in the CI workflow ?

- The documentation/website structure is difficult to keep up to date, with a submodule in the middle, a bash script to update it. This is not clean and needs to be improved.

- The documentation is written for the website only (my mistake): By using the documentation within the website, some formatting and browsing has been written only to work with the website. I've started to rewrite the website/documentation few weeks ago and I would like to get this done. We can discuss the changes and work in this issue (or create a new one).

- Website is not tracking the latest version of airsonic dynamically, it needs to be update by hand, and it's annoying.

I might have missed a few points here but I'd like to get people's thoughts about this. Maybe also say what process that I missed could be automated and/or improved.

:beers: Cheers

**Comment by jooola** [comment]
This issue is part of the release plan for v11.0 https://github.com/airsonic/airsonic/issues/1439

**Comment by randomnicode** [comment]
Some of these points (particularly related to CI/CD release automation and changelog generation) were already brought up and a PR was even submitted to that effect, but the proposal was explicitly vetoed. The subsequent successors to that PR were merged on the `airsonic-advanced` fork.

Maintainers seem to prefer releasing by hand. They think it's safer. And to release less frequently.

Release automation is actually one of the reasons `airsonic-advanced` was forked.

**Comment by randomnicode** [comment]
Additionally, while I think it's fine to use a milestone marker for CI/CD improvements (just to be able to track progress), the release versioning should really have no bearing on release cycles.

You could make changes to CI/CD today without having 11.0 timelines hanging in the balance just because the release process isn't ready. All that will accomplish is that it will extend the 11.0 timelines even farther.

**Comment by tesshucom** [comment]
Translation is ... 
Though it seems to be the easiest to automate.
I don't know exactly what's going on because the mechanism changes and there is no announcement for third party participants.
I would be happy if there was a simple mechanism that even non-developers could participate.

**Comment by tesshucom** [comment]
> Current testing in Travis seems to be full of false positive, or really unstable. I need some insight on this one, has been a while I've not been around. (Also this could have a separate issue)

As far as I remember, Airsonic faced Travis a major obstacle only once.
As Travis discloses, there are other minor failur, but they are not a problem.
In other words, Travis is stable at a practical level.

Another issue, as you pointed out.
It is about the human rules of how to respond to CVE, and changing the CI system does not that it can be improved.

I ran weekly clone on the my clone and almost never got red.
This is due to self-correction within one week.
That's what you can do because it's a clone.

If we want to balance responsiveness and risk in co-management, we may need an idea like nightly, as suggested once by randomnicode.

**Comment by fxthomas** [comment]
> Has checkstyle been integrated in the maven tests ? It this already integrated in the CI workflow ?

I do believe so, I occasionally saw it help me write better code ;)

**Comment by randomnicode** [comment]
> I do believe so, I occasionally saw it help me write better code ;)

I'd love for you to expand on this statement. Do you have any particular examples of code it's made you write better?

**Comment by tesshucom** [comment]
There is no perfect car in our world, and what exists is a reasonable shipping standard.
Although checkstyle can be one of the indicators, it is not the most important standard for quality assurance.
On the contrary improper adoption and fanaticism reduces productivity.

Putting that aside...
It should be noted that jooola offers a wide variety of very good and important perspectives.
If we want to organize very objectively, it may be better to categorize a little.
Or we can make a supplement to the classification.

For example, "what users long for today" or "what will help developers in the medium to long term".
These criteria are self-explanatory and both are important.

Priorities can be set within each group and reflected in plan based on those priorities well-balanced.
If we can see Airsonic's policies and priorities are clear, we can attract the attention of collaborators.

**Comment by tesshucom** [comment]
There is no perfect car in our world, and what exists is a reasonable shipping standard.
Although checkstyle can be one of the indicators, it is not the most important standard for quality assurance.
On the contrary improper adoption and fanaticism reduces productivity.

Putting that aside...
It should be noted that jooola offers a wide variety of very good and important perspectives.
We may need to do some tidying up for easier tracking.

Well, we probably need issues for each...

Category | Title | Sub issues | MileStone
-- | -- | -- | --
Engineer | Automation of CI/CD workflow | #1561 | -
User/Engineer | Introduction of Angular commit message format (automation of change log) | #1561 | -
Engineer | A study on improving the operation of Travis (CI) | #1561 | -
User/Engineer | Proposal for doc/web update process | #1561 | -
User/Engineer | Suggestions for website version updates | #1561 | -
User/Engineer | Suggestions for improvement of doc description method | #1561 | -
User | Suggestions for improving translation automation | #1561 | -

**Comment by fxthomas** [comment]
Yay, I'm glad we're picking up this subject, especially given the increased number of contributors.

The nightly build was discussed in #1373, @tesshucom do you mind if I edit your post directly to add it? If someone wants to pick it up and handle the whole thing, let me know ; otherwise I intended to work on this once most of the things are settled for 11.0.

----

> I'd love for you to expand on this statement. Do you have any particular examples of code it's made you write better?

Mostly code style like braces, import order, spaces etc. Having a uniform codebase is a lot better when you're reading code, and Airsonic is doing pretty well for that (at least compared to some of the legacy stuff I sometimes see elsewhere :wink:).

>  We may need to do some tidying up for easier tracking.

:heart: Good thinking!

> Current testing in Travis seems to be full of false positive, or really unstable. I need some insight on this one, has been a while I've not been around.

Travis itself seems fairly okay for me, but some parts of the test process are more brittle. Namely, Docker integration tests. I'm not at all familiar with them though, so I've mostly left them untouched and don't even run them on a regular basis on my machine.

> I started to rewrite the website/documentation few weeks ago and I would like to get this done. We can discuss the changes and work in this issue (or create a new one).

@jooola Do create a new one! I like to keep separate topics in separate issues, makes it easier to take decisions on each separately.

> Website is not tracking the latest version of airsonic dynamically, it needs to be update by hand, and it's annoying.

The manual process is not even documented in https://github.com/airsonic/airsonic/blob/master/contrib/release.md, that's definitely a low-hanging fruit that we could add with a simple PR.

**Comment by tesshucom** [comment]
> do you mind if I edit your post directly to add it?

Yes, please edit directly.
There are a lot of sub issues hanging under this issue-(Work on a better workflow).

You might want to replace Jooola's original proposal with a table.
Derived issues probably have different lifecycles.
Some resolve relatively quickly and some require long-term struggle.

If it can be subdivided, it can be considered that this issues itself has ended its role as a rocket launcher.

**Comment by jooola** [comment]
I added the table in my first post. I'll open sub issues and reference them in here.

Discussion will go further on the dedicated issues.
